### PR TITLE
InitOS extensions

### PR DESCRIPTION
**Pull request by codingforfun**
Collection of fixes and extensions we made.

### CODE REVIEW

**Review by initBasti — CHANGES_REQUESTED** [review]
This was an error, please dismiss this request.

**Review by initBasti — APPROVED** [review]
I like this change a lot, it drops some unnecessary if clauses and is more descriptive than the previous version.

**Comment by ortlam** [comment]
@initBasti I can remove the commented-out code if you want. It was left in because we were not sure whether removing the check was a permanent or temporary thing. If you feel like it served a useful purpose in avoiding user error, maybe allowing to opt-out would be better. What do you think?

File: plenty_api/api.py
Comment (initBasti) [inline plenty_api/api.py:776]: I believe this can be handled as well with the `plenty_api_generic_get` wrapper. Was there something particular that stopped you from using it?

File: plenty_api/constants.py
Comment (initBasti) [inline plenty_api/constants.py:49]: This is nice! thank you.

**Review by initBasti — COMMENTED** [review]
Overall these get methods look like a great addition, I will write the documentation for them right after the review. I have added 2 questions below.

File: plenty_api/api.py
Comment (initBasti) [inline plenty_api/api.py:334]: Was this an error?

File: plenty_api/api.py
Comment (codingforfun) [reply plenty_api/api.py:747]: To be honest I’m not sure what I originally meant here anymore. I think it was about removing keys from the returned data, because `plenty_api_get_storagelocations` returns more data than necessary here.

File: plenty_api/api.py
Comment (initBasti) [inline plenty_api/api.py:648]: What is the purpose of this query dictionary? You should adjust the documentation of a public method if you change its parameter set. The whole idea behind `refine` and `additional`, was to abstract the raw query into something more user-friendly. Is this some kind of alternative to that idea? 
Additionally, I find it strange that you introduce that new parameter for the `get` variations and items methods but not for manufacturers where you keep on working with the local query dictionary. This is again a reason why I think that commit descriptions are super useful as I am not sure if this solves a particular use case or if manufacturers should be adjusted as well?

File: plenty_api/api.py
Comment (initBasti) [inline plenty_api/api.py:310]: just for your information, I will drop this commit during merge as I already have something similar on my tree.

File: plenty_api/api.py
Comment (initBasti) [inline plenty_api/api.py:832]: I agree that this method is not optimal but this change feels out of place. Especially because the other return paths still return a boolean and the method description still declares the return value to be a boolean.

File: plenty_api/api.py
Comment (initBasti) [inline plenty_api/api.py:854]: I believe the `bestBeforeDate` parameter should be parsed first by the `utils.parse_date` function, which makes that a) the string contains a date and b) that it is in the W3C format.
something along the lines of:
```
if bestBeforeDate:
    w3c_date = utils.parse_date(date=bestBeforeDate)
    if w3c_date:
        data.update(bestBeforeDate=w3c_date)
```

File: plenty_api/api.py
Comment (initBasti) [inline plenty_api/api.py:886]: same comment as on line 850

File: plenty_api/utils.py
Comment (ortlam) [reply plenty_api/utils.py:304]: We are working with a server which would not fit any pattern beyond `https://.*`. I am not familiar enough with PlentyMarkets to know if this is a service they offer or an entirely different server. The custom URL is set internally in the PlentyMarkets server and listed as REST endpoint. I believe this means it's impossible to check the URL.

File: plenty_api/api.py
Comment (initBasti) [inline plenty_api/api.py:836]: The naming convention of the function can be open for discussion but for now, I would fix those names during the merge in order to align with the rest of the API interface. 
Meaning:
s/book_incoming_items/plenty_api_book_incoming_items/
s/book_outgoing_items/plenty_api_book_outgoing_items/

My reason for this convention was for the methods to be as descriptive as possible while being used within a user application. (Especially making clear when the plenty API is touched)
I would rather say, if we agree on a new naming convention for the methods then we do it for all the public methods for v.0.2.8.

File: plenty_api/api.py
Comment (initBasti) [inline plenty_api/api.py:167]: That looks like a copy-paste accident

File: plenty_api/api.py
Comment (initBasti) [inline plenty_api/api.py:148]: I would like to get a view of the use case here. Does this mean that you store the plain password within a file/configuration, read it out within the user application, and put it into the init? If that is the case, what exactly didn't you like about the gpg file/ keyring method? Is it the authentication? I simply ask because I don't want to encourage users of the API interface to use unsafe methods for storing their passwords. Especially, because an attacker can do quite a bit of financial damage when he gets access to the REST API account of a plentymarkets system.

**Review by initBasti — COMMENTED** [review]
And here the second part of the review, as soon as we have resolved the last remaining issues I will merge the whole pull request.
Thank you :)

File: plenty_api/utils.py
Comment (initBasti) [reply plenty_api/utils.py:304]: Okay thank you, we should drop it in this case.

File: plenty_api/api.py
Comment (initBasti) [reply plenty_api/api.py:747]: Okay, then I would say we drop the comment.

File: plenty_api/api.py
Comment (ortlam) [reply plenty_api/api.py:832]: This is related to a change in `__plenty_api_request` which returns the error response instead of `None` when a request fails. I changed this because I need the error message to forward it to the end user, but the focus was on getting it to work quickly.
What do you think is the best way we should be forwarding error messages? Exceptions might be cleaner, but would require more code to catch them when using the API.

File: plenty_api/api.py
Comment (ortlam) [reply plenty_api/api.py:148]: I think it should be up to the user of the library to make sure whatever way to store credentials is safe. Encrypted files and keyrings are good ways to do this, but you might also get them from a database, from some window or from the terminal (which you already implemented). This change allows the program implementing your library to provide a password from an alternative source without any workarounds.

File: plenty_api/api.py
Comment (initBasti) [reply plenty_api/api.py:832]: I guess the best course of action, is to rewrite that method as it doesn't align with the rest of the library. I will do that after merging this patchset and before publishing v.0.2.8

File: plenty_api/api.py
Comment (ortlam) [reply plenty_api/api.py]: That was an oversight, I fixed it now.

File: plenty_api/api.py
Comment (ortlam) [reply plenty_api/api.py]: Yeah, should be fixed.

File: plenty_api/api.py
Comment (ortlam) [reply plenty_api/api.py]: This shouldn't have been in the PR, I removed it.

File: plenty_api/api.py
Comment (ortlam) [reply plenty_api/api.py]: Good call, thank you.

File: plenty_api/api.py
Comment (ortlam) [reply plenty_api/api.py:832]: You are right, we should have named them according to the convention. Would you prefer me to fix it or to do it yourself?

File: plenty_api/api.py
Comment (ortlam) [reply plenty_api/api.py:167]: It was, I fixed the data type.

File: plenty_api/api.py
Comment (ortlam) [reply plenty_api/api.py]: I have renamed them now.

**Comment by initBasti** [comment]
Thank you for the pull request, the changes have been merged into master. It is part of the newly published v.0.2.8 (https://pypi.org/project/plenty-api/). A few remarks:
- In order to align the term to the rest of the codebase, I renamed `article_item_id` to `item_id`, please make sure to check that in your application
- In the commit you will be able to see a few fix-ups, mostly these contain fixes off unused parameter variables, missing valid values, invalid coding style, fixing pylint warnings, make the generic get method private, adding some commit description if necessary.
- I also added documentation for the new methods

**Pull request closed without merging**
